### PR TITLE
enhance: avoid context limit

### PR DESCRIPTION
**Pull request by g-linville**
for gptscript-ai/desktop#283

This PR introduces three new measures to try to help the user avoid hitting the context length error:

1. Garbage collection works now. Previously, the token threshold was set too high, so we would run out of context before garbage collection would ever hit. Now, older messages will get deleted from the chat (on the LLM's side only; users can still see them) once it is needed in order to stay under the context limit.
2. If the output of a single tool output surpasses approximately 80% of the total context window, the tool output is considered too long and an error message is given to the LLM. We will design tools with the context limits in mind, but this is a safeguard to prevent just crashing and to allow the user to continue chatting with the LLM, and to allow the LLM to adjust its tool arguments and possibly remain under the output limit.
3. If, despite the previous measures, we do somehow hit the context length limit error, we catch that error and try to garbage collect more aggressively, and then retry the call up to ten times, doing more aggressive garbage collection each time.

With these three measures, I am fairly confident that users will no longer encounter the error.

### CODE REVIEW

**Review by ibuildthecloud — APPROVED** [review]
The very important thing is that we never drop the system role message. I believe that is still that case.

**Comment by g-linville** [comment]
@ibuildthecloud correct. The way I implemented this will never change how the system role messages are handled. Those will always stick around.

File: pkg/openai/client.go [hunk 1]
```diff
@@ -317,6 +320,14 @@ func (c *Client) Call(ctx context.Context, messageRequest types.CompletionReques
 	}
 
 	if messageRequest.Chat {
+		// Check the last message. If it is from a tool call, and if it takes up more than 80% of the budget on its own, reject it.
+		lastMessage := msgs[len(msgs)-1]
+		if lastMessage.Role == string(types.CompletionMessageRoleTypeTool) && countMessage(lastMessage) > int(math.Round(float64(getBudget(messageRequest.MaxTokens))*0.8)) {
```
Comment (thedadams) [inline pkg/openai/client.go:325]: nit: probably don't need to `math.Round` here. Not much of a difference between 102,399 and 102,400.

File: pkg/openai/client.go [hunk 2]
```diff
@@ -383,6 +394,16 @@ func (c *Client) Call(ctx context.Context, messageRequest types.CompletionReques
 		return nil, err
 	} else if !ok {
 		response, err = c.call(ctx, request, id, status)
+
+		// If we got back a context length exceeded error, keep retrying and shrinking the message history until we pass.
+		var apiError *openai.APIError
+		if err != nil && errors.As(err, &apiError) && apiError.Code == "context_length_exceeded" && messageRequest.Chat {
```
Comment (thedadams) [inline pkg/openai/client.go:400]: nit: `errors.As` takes care of the `err != nil` check
```suggestion
		if errors.As(err, &apiError) && apiError.Code == "context_length_exceeded" && messageRequest.Chat {
```

File: pkg/openai/client.go [hunk 3]
```diff
@@ -421,6 +442,32 @@ func (c *Client) Call(ctx context.Context, messageRequest types.CompletionReques
 	return &result, nil
 }
 
+func (c *Client) contextLimitRetryLoop(ctx context.Context, request openai.ChatCompletionRequest, id string, maxTokens int, status chan<- types.CompletionStatus) ([]openai.ChatCompletionStreamResponse, error) {
+	var (
+		response []openai.ChatCompletionStreamResponse
+		err      error
+	)
+
+	for range 10 { // maximum 10 tries
```
Comment (thedadams) [inline pkg/openai/client.go:451]: Is this the first use in our code base?

File: pkg/openai/count.go [hunk 4]
```diff
@@ -1,20 +1,32 @@
...
+	"math"
+
+	openai "github.com/gptscript-ai/chat-completion-client"
+)
+
+const DefaultMaxTokens = 128_000
+
+func decreaseTenPercent(maxTokens int) int {
+	maxTokens = getBudget(maxTokens)
+	return int(math.Round(float64(maxTokens) * 0.9))
```
Comment (thedadams) [inline pkg/openai/count.go:13]: same nit about `math.Round`

**Review by thedadams — APPROVED** [review]

File: pkg/openai/client.go (hunk 3 above)
Comment (g-linville) [reply pkg/openai/client.go:451]: I think so!

File: pkg/openai/client.go (hunk 1 above)
Comment (g-linville) [reply pkg/openai/client.go]: Fixed

File: pkg/openai/client.go (hunk 2 above)
Comment (g-linville) [reply pkg/openai/client.go]: Fixed

File: pkg/openai/count.go (hunk 4 above)
Comment (g-linville) [reply pkg/openai/count.go]: Fixed

**Pull request merged**